### PR TITLE
Native CodeMirror 6 support

### PR DESCRIPTION
**Pull request by personalizedrefrigerator**
# Summary

At present, the Math Mode plugin doesn't work well in the CodeMirror 6-based beta editor. This pull request refactors the CodeMirror plugin to:
1. Use CodeMirror 6 APIs when running in CodeMirror 6
2. Use CodeMirror 5 APIs when running in CodeMirror 

Previously, CodeMirror 6 support was provided by Joplin's compatibility layer.

# Why avoid the compatibility layer?

- Bugs
    - Currently, only the last line of math seems to be rendered correctly.
    - Issues with selection and inline results.
- The compatibility layer may be removed in the future.

# Screen recording

https://github.com/user-attachments/assets/5181f709-76c2-4a51-bdcc-412ec852cd9a

# Notes and resources

- [Tutorial: Creating a Joplin plugin compatible with both CodeMirror 5 and CodeMirror 6](https://joplinapp.org/help/api/tutorials/cm6_plugin#codemirror-5-compatibility) (takes a slightly different approach)
- [CodeMirror 6 API documentation](https://codemirror.net/)
- To allow correctly importing CodeMirror 6 libraries, the plugin build script was [updated](https://joplinapp.org/help/api/tutorials/cm6_plugin#update-the-plugin-build-script) to the latest version.
   - See the changes to `webpack.config.js` and some of the changes to `package.json`.
- Many of the lines marked as added by `git` were moved from `src/mathMode.ts`, which was split into separate files.

# To-do

- [x] Improve refresh logic — currently, math isn't refreshed in some cases:
    - [x] Adding a configuration line to a math block.
    - [x] Converting a math block to a different type of code block.
- [x] Remove changes to `package-lock.json` from this pull request — such changes are often difficult to review and can be made separately.
- [x] Refactoring, code cleanup, and documentation
    - **Optional?**: Consider taking the approach outlined in the [CodeMirror 5 compatibility documentation](https://joplinapp.org/help/api/tutorials/cm6_plugin#codemirror-5-compatibility), where the CodeMirror 5 and CodeMirror 6 content scripts are bundled separately by Webpack. At present, CodeMirror 6 and CodeMirror 5 use the same entrypoint.
- [x] More manual testing.

### CODE REVIEW

File: package.json
Comment (personalizedrefrigerator) [inline package.json:1]: Summary of changes to `package.json`:
- Updated to the [latest template version](https://github.com/laurent22/joplin/blob/dev/packages/generator-joplin/generators/app/templates/package_TEMPLATE.json).
- Added a new dev dependency:
  ```js
  "@codemirror/view": "^6.26.3",
  ```

**Comment by CalebJohn** [comment]
Thanks for these updates @personalizedrefrigerator! I noticed that there seems to be a few bugs using this update however. Please take a look below

Just to note, I don't care too much about the Codemirror 5 behavior, as it won't be the default for long, but I would like to get the Codemirror 6 behavior solid before release.

Thanks again!

# Before (for reference)
image
image



# After - Codemirror 5
We are seeing output next to the config lines after this update
image
Looks like the magic "total" variable is no longer working after this change.
image


# After - Codemirror 6
This note looks great!
image
Same issue with the total variable
image

## After visiting a note with `displaytotal: yes`
This issue is more worrying, after visiting a note with the "displaytotal" config, it looks like the configuration is accidentally stored globally and not cleared when switching notes. This breaks the note-specific config blocks.
image


<details><summary>Note 1</summary>
<p>

```math
global: yes
displaytotal: yes

1
2
3
```



</p>
</details> 

<details><summary>Note 2</summary>
<p>

```math
align: right

Base = 832000 USD
Extra = 1200000 USD
total
total in CAD
```

</p>
</details>

**Comment by personalizedrefrigerator** [comment]
> Thanks for these updates @personalizedrefrigerator! I noticed that there seems to be a few bugs using this update however. Please take a look below

All three issues should now be fixed!

- CodeMirror 5: Config results shown as "undefined". ([Commit](https://github.com/CalebJohn/joplin-math-mode/pull/26/commits/997ac9c76fcede2ac75b7abc254dae1f4cfbe557)).
    - This was caused by a refactoring that was only partially applied. See the commit message for additional details.
- The `total` variable was always `""`.
    - This was caused by setting the default `scope` to `{ total: '' }`. This then overwrote the `{ total: block_total }` definition of the `total` variable.
- Global configuration was preserved after switching notes ([commit](https://github.com/CalebJohn/joplin-math-mode/pull/26/commits/1e01a206e5d1469aa3ebc8c912678abe85f713af)):
   - Fixed by replacing an `Object.assign` with an `{ ... }`. `Object.assign` modifies its first argument. In this case, the first argument (`globalConfig`) should not have been modified.

**Comment by CalebJohn** [comment]
Thanks for this @personalizedrefrigerator! Everything looks great here, I'm looking forward to the editor switch!

**Comment by personalizedrefrigerator** [comment]
I'm linking to the [upstream pull request](https://github.com/laurent22/joplin/pull/11143) that removes the "incompatible" warning in the new editor for Math Mode.

**Pull request merged**